<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
مشکل مسکن
</font>
</h1>

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
مقدمه و صورت مسئله
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
در این سوال ما به دنبال آموزش یک مدل یادگیری ماشین هستیم که بتوان با کمک آن قیمت مسکن در کشور پهناور ایران را پیش‌بینی کنیم. با استفاده از مجموعه داده‌ای که در اختیار شما قرار گرفته است، به ایجاد و تخمین متغیری به نام <code>Total</code> می‌پردازیم. این متغیر در مجموعه‌داده‌ ما وجود ندارد. در ادامه به نحوه ایجاد و محاسبه‌ی آن پرداخته می‌شود.
</font>
</p>

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
معرفی مجموعه داده
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
در فایل اولیه‌ی این سوال، یک پوشه با نام <code>Data</code> وجود دارد که شامل ۲ فایل <code>train.csv</code> و <code>test.csv</code> وجود دارد. در این بخش به بررسی این ۲ فایل می‌پردازیم.
</font>
</p>


<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
هرستون این مجموعه‌داده با توجه به ماهیت خود می‌تواند مقادیر عددی یا رشته‌ای داشته باشند. برای آشنایی بیشتر با مجموعه‌داده، پیشنهاد می‌شود هرستون به صورت جداگانه بررسی شود.</font>
</p>


<h4 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
معرفی مجموعه داده آموزشی (train.csv)
</font>
</h4>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
فایل <code>train.csv</code> دارای 21 ستون می‌باشد که ستون‌های آن به شرح زیر می‌باشد:
<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>Area</code>|متراژ خانه‌های به متر مربع|
|<code>Year_built</code>|سال ساخت ساختمان|
|<code>Rooms</code>|تعداد اتاق‌ها|
|<code>Floor</code>|شماره‌ی طبقه‌ی واحد در خانه|
|<code>Elevator</code>|وجود آسانسور|
|<code>Parking</code>|وجود پارکینگ|
|<code>StorageRoom</code>|وجود انباری|
|<code>Rent_type</code>|نوع اجاره‌ی خانه|
|<code>Units_per_floor</code>|تعداد واحد در هر طبقه|
|<code>Building_direction</code>|جهت جغرافیایی ساختمان|
|<code>Unit_status</code>|وضعیت بازسازی واحد|
|<code>Floor_material</code>|جنس کف‌پوش|
|<code>Bathroom</code>|نوع دستشویی واحد|
|<code>Balcony</code>|وجود بالکن|
|<code>Heating</code>|نوع گرمایش|
|<code>Cooling</code>|نوع سرمایش|
|<code>Hot_water_provider</code>|نوع گرم‌کن آب|
|<code>District</code>|منطقه جغرافیایی در شهر|
|<code>City</code>|شهری که واحد مسکونی در آن واقع شده|
|<code>Credit</code>|میزان پول پیش برای اجاره‌ی واحد (میلیون تومان)|
|<code>Rent</code>|میزان اجاره واحد (میلیون تومان)|

</font>
</div>
</center>

<h4 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
معرفی مجموعه داده آزمون (test.csv)
</font>
</h4>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
فایل <code>test.csv</code> دارای ۱۹ ستون می‌باشد که ستون‌های آن به شرح زیر می‌باشد:


<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>Area</code>|متراژ خانه‌های به متر مربع|
|<code>Year_built</code>|سال ساخت ساختمان|
|<code>Rooms</code>|تعداد اتاق‌ها|
|<code>Floor</code>|شماره‌ی طبقه‌ی واحد در خانه|
|<code>Elevator</code>|وجود آسانسور|
|<code>Parking</code>|وجود پارکینگ|
|<code>StorageRoom</code>|وجود انباری|
|<code>Rent_type</code>|نوع اجاره‌ی خانه|
|<code>Units_per_floor</code>|تعداد واحد در هر طبقه|
|<code>Building_direction</code>|جهت جغرافیایی ساختمان|
|<code>Unit_status</code>|وضعیت بازسازی واحد|
|<code>Floor_material</code>|جنس کف‌پوش|
|<code>Bathroom</code>|نوع دستشویی واحد|
|<code>Balcony</code>|وجود بالکن|
|<code>Heating</code>|نوع گرمایش|
|<code>Cooling</code>|نوع سرمایش|
|<code>Hot_water_provider</code>|نوع گرم‌کن آب|
|<code>District</code>|منطقه جغرافیایی در شهر|
|<code>City</code>|شهری که واحد مسکونی در آن واقع شده|

</font>
</div>
</center>


<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
در این قسمت کتاب‌خانه‌ها و ابزار مورد نیاز خود را <code>import</code> کنید و فایل داده‌ها را که در پوشه‌ی <code>Data</code> و که با نام‌های <code>train.csv</code> و <code>test.csv</code> ذخیره‌شده‌اند را بخوانید و وارد محیط کار خود کنید.
</font>
</p>

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.preprocessing import RobustScaler
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestRegressor
import warnings
warnings.filterwarnings('ignore')


df = pd.read_csv('data/train.csv')
test = pd.read_csv('data/test.csv')


df.head()

,Area,Year_built,Rooms,Floor,Elevator,Parking,StorageRoom,Rent_type,Units_per_floor,Building_direction,...,Floor_material,Bathroom,Balcony,Heating,Cooling,Hot_water_provider,District,Credit,Rent,city
0,103.0,1390,2,2,1.0,1.0,1.0,غیر قابل تبدیل,انتخاب نشده,انتخاب نشده,...,انتخاب نشده,انتخاب نشده,دارد,شوفاژ,اسپلیت,پکیج,ansari,500.0,11.0,rasht
1,85.0,1394,2,2 از 6,1.0,1.0,0.0,غیر قابل تبدیل,4,جنوبی,...,سرامیک,ایرانی,دارد,انتخاب نشده,کولر گازی,آبگرم‌کن,unknown,450.0,0.0,bandar-abbas
2,90.0,1402,2,4 از 5,1.0,1.0,0.0,NaN,4,جنوبی,...,سرامیک,ایرانی و فرنگی,ندارد,شوفاژ,کولر آبی,پکیج,unknown,170.0,1.5,parand-city
3,160.0,1399,3,1 از 3,1.0,1.0,1.0,غیر قابل تبدیل,1,جنوبی,...,سرامیک,ایرانی و فرنگی,دارد,شوفاژ,اسپلیت,پکیج,sattarkhan-shiraz,1200.0,6.0,shiraz
4,78.0,1396,2,14,1.0,1.0,0.0,غیر قابل تبدیل,انتخاب نشده,انتخاب نشده,...,انتخاب نشده,انتخاب نشده,انتخاب نشده,انتخاب نشده,انتخاب نشده,انتخاب نشده,unknown,100.0,4.0,tabriz


<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
ایجاد متغیر هدف
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
در این قسمت باید متغیر هدف مسئله را ایجاد کنیم. در این مسئله هدف ما تخمین ارزش خانه‌های مورد نظر است. برای این منظور، با استفاده از فرمول زیر، ستونی با نام <code>Total</code> به مجموعه داده خود اضافه کنید تا مدل یادگیری ماشین خود را منطبق بر آن آموزش دهید.

$$ Total = Credit + (Rent * 33.33334)$$

</font>
</p>

In [21]:

df['Total'] = df['Credit'] + (df['Rent'] * 33.33334)



delete outlier

In [22]:
lower_bound = df['Total'].quantile(0.01)
upper_bound = df['Total'].quantile(0.99)
df_clean = df[(df['Total'] >= lower_bound) & (df['Total'] <= upper_bound)]

# Log transform target 
df_clean['Total_log'] = np.log1p(df_clean['Total'])

print(f"\nLog transformed target created")


Log transformed target created


cleaning data

In [ ]:
def engineer_features(data, is_train=True, encoders=None):
   
    df_feat = data.copy()
    
    
    numeric_cols = ['Area', 'Year_built', 'Rooms', 'Units_per_floor']
    for col in numeric_cols:
        if col in df_feat.columns:
            df_feat[col] = pd.to_numeric(df_feat[col], errors='coerce')
    
   
    if 'Floor' in df_feat.columns:
        df_feat['Floor_clean'] = df_feat['Floor'].astype(str)
        
        df_feat['Floor_Num'] = df_feat['Floor_clean'].str.extract(r'(\d+)')[0]
        df_feat['Floor_Num'] = pd.to_numeric(df_feat['Floor_Num'], errors='coerce')
      
        df_feat['Floor_Num'] = df_feat['Floor_clean'].apply(
            lambda x: 0 if 'همکف' in str(x) else x
        )
        df_feat['Floor_Num'] = pd.to_numeric(df_feat['Floor_Num'], errors='coerce')
        
        df_feat['Floor_Total'] = df_feat['Floor_clean'].str.extract(r'از\s*(\d+)')[0]
        df_feat['Floor_Total'] = pd.to_numeric(df_feat['Floor_Total'], errors='coerce')
    
    binary_cols = ['Elevator', 'Parking', 'StorageRoom', 'Balcony']
    for col in binary_cols:
        if col in df_feat.columns:
            df_feat[col] = df_feat[col].astype(str)
            df_feat[col] = df_feat[col].map({'دارد': 1, 'ندارد': 0})
            df_feat[col] = df_feat[col].fillna(0)
    

    if 'Year_built' in df_feat.columns:
        current_year = 2026
        df_feat['Building_Age'] = current_year - df_feat['Year_built']
        df_feat['Building_Age'] = df_feat['Building_Age'].clip(0, 100)
        df_feat['Is_New'] = (df_feat['Building_Age'] <= 5).astype(int)
        df_feat['Is_Old'] = (df_feat['Building_Age'] >= 30).astype(int)
    
   
    if 'Area' in df_feat.columns and 'Rooms' in df_feat.columns:
        df_feat['Area_per_Room'] = df_feat['Area'] / df_feat['Rooms']
        df_feat['Area_per_Room'] = df_feat['Area_per_Room'].replace([np.inf, -np.inf], np.nan)
    
   
    if 'Floor_Num' in df_feat.columns and 'Floor_Total' in df_feat.columns:
        df_feat['Floor_Ratio'] = df_feat['Floor_Num'] / df_feat['Floor_Total']
        df_feat['Floor_Ratio'] = df_feat['Floor_Ratio'].replace([np.inf, -np.inf], np.nan)
        df_feat['Is_Ground'] = (df_feat['Floor_Num'] == 0).astype(int)
        df_feat['Is_Top'] = ((df_feat['Floor_Num'] == df_feat['Floor_Total']) & 
                             (df_feat['Floor_Total'].notna())).astype(int)
    

    if 'Area' in df_feat.columns:
        df_feat['Area_Squared'] = df_feat['Area'] ** 2
        if 'Elevator' in df_feat.columns:
            df_feat['Area_Elevator'] = df_feat['Area'] * df_feat['Elevator']
        if 'Parking' in df_feat.columns:
            df_feat['Area_Parking'] = df_feat['Area'] * df_feat['Parking']
    
    
    cat_cols = ['Rent_type', 'Building_direction', 'Unit_status', 
                'Floor_material', 'Bathroom', 'Heating', 'Cooling', 
                'Hot_water_provider', 'District', 'city']
    
    for col in cat_cols:
        if col in df_feat.columns:
            if is_train:
          
                freq_map = df_feat[col].value_counts(normalize=True).to_dict()
                df_feat[col + '_enc'] = df_feat[col].map(freq_map)
                if encoders is not None:
                    encoders[col] = freq_map
            else:
              
                if encoders and col in encoders:
                    df_feat[col + '_enc'] = df_feat[col].map(encoders[col])
                    df_feat[col + '_enc'] = df_feat[col + '_enc'].fillna(0)
    
    return df_feat, encoders

In [24]:

encoders = {}
df_engineered, encoders = engineer_features(df_clean, is_train=True, encoders=encoders)


test_engineered, _ = engineer_features(test, is_train=False, encoders=encoders)


In [35]:

feature_cols = [
    'Area', 'Rooms', 'Units_per_floor',
    'Building_Age', 'Is_New', 'Is_Old',
    'Elevator', 'Parking', 'StorageRoom', 'Balcony',
    'Area_per_Room', 'Area_Squared', 'Area_Elevator', 'Area_Parking',
    'Floor_Num', 'Floor_Total', 'Floor_Ratio', 'Is_Ground', 'Is_Top'
]

encoded_cols = [col for col in df_engineered.columns if col.endswith('_enc')]
feature_cols.extend(encoded_cols)


feature_cols = [col for col in feature_cols if col in df_engineered.columns]




In [26]:

X = df_engineered[feature_cols].copy()
y = df_engineered['Total_log'].copy()



In [27]:

imputer = SimpleImputer(strategy='median')
X_imputed = pd.DataFrame(imputer.fit_transform(X), columns=feature_cols)
test_imputed = pd.DataFrame(imputer.transform(test_engineered[feature_cols]), columns=feature_cols)


In [ ]:

X_train, X_val, y_train, y_val = train_test_split(
    X_imputed, y, test_size=0.2, random_state=42
)



In [29]:

scaler = RobustScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)
test_scaled = scaler.transform(test_imputed)


In [30]:


rf = RandomForestRegressor(
    n_estimators=300,      # Number of trees
    max_depth=20,          # Maximum depth of each tree
    min_samples_split=10,  # Minimum samples to split a node
    min_samples_leaf=5,    # Minimum samples in a leaf node
    random_state=42,       # For reproducibility
    n_jobs=-1              # Use all CPU cores
)

rf.fit(X_train_scaled, y_train)
print("training completed")

training completed


r2 score 

In [ ]:

y_val_pred_log = rf.predict(X_val_scaled)

#back from log to original
y_val_pred = np.expm1(y_val_pred_log)
y_val_actual = np.expm1(y_val)


r2 = r2_score(y_val_actual, y_val_pred)
rmse = np.sqrt(mean_squared_error(y_val_actual, y_val_pred))


print(f"R² Score: {r2:.4f}    ")



R² Score: 0.8138    


<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
<b>ارسال‌ نهایی</b>
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    کد خود را به‌همراه    یک گزارش‌کار از نحوه عملکرد کد خود و نحوه پیش‌پردازش و تکنیک‌های خود را در یک فایل زیپ ارسال کنی
</font>

This project develops a machine learning model to predict housing prices in Iran using a dataset of 382,467 training records. Key steps include creating the target variable Total = Credit + (Rent × 33.33334), applying log transformation to handle skewed price distribution, and engineering 20+ features such as Building_Age, Area_per_Room, and Floor_Ratio. Target encoding with smoothing was used for categorical variables like city and district. A Random Forest Regressor achieved an R² score of 0.81 on validation data, with area and location identified as the most important price predictors.